content/blog/trungnt2910/2024-08-11_[gsoc_2024]_debugging_progress_3.md

```diff
@@ -0,0 +1,370 @@
+GDB has been modified to link to this. However, for the compile macro to work, we would need to add
+`-DB_USE_POSITIVE_POSIX_ERRORS` to the preprocessor flags. Adding it to just GDB's `Makefile`s will
+not suffice since some `gdbsupport` code is also statically compiled into GDB, and these files are
+apparently not affected by GDB's `CFLAGS`.
+
+The recipe has been updated to handle this detail. For source builds, you would need to re-build
+everything from a clean directory, and run `export CPPFLAGS="-DB_USE_POSITIVE_POSIX_ERRORS"` before
+executing the `./configure` script.
+
+## Debugging C++ Types
+
+A common task while debugging C++ codebases is inspecting STL containers.
+
+While the GDB that comes with popular Linux distros seems to have support for that built-in, we do
+not get that for free in our Haiku port:
+
+```
+(gdb) thread 6
+[Switching to thread 6 (team 371 (gdb) thread 381 (gdb debugger port listener))]
+#0  0x000001302b72828c in _kern_wait_for_objects () from /boot/system/lib/libroot.so
+(gdb) up 1
+#1  0x0000020f32e269b7 in operator() (__closure=0x0, data=0x0) at ../../gdb-haiku/gdb/nat/haiku-nat.c:2132
+2132                        event_count = wait_for_objects (wait_infos.data (),
+(gdb) info locals
+event_count = 0
+wait_infos = {<std::_Vector_base<object_wait_info, std::allocator<object_wait_info> >> = {
+    _M_impl = {<std::allocator<object_wait_info>> = {<std::__new_allocator<object_wait_info>> = {<No data fields>}, <No data fields>}, <std::_Vector_base<object_wait_info, std::allocator<object_wait_info> >::_Vector_impl_data> = {_M_start = 0x10efbc963c90,
+        _M_finish = 0x10efbc963ca0, _M_end_of_storage = 0x10efbc963ca0}, <No data fields>}}, <No data fields>}
+wait_for_ports = true
+__func__ = "operator()"
+(gdb)
+```
+
+The "pretty printers" for types living in `libstdcxx` does not come with GDB. It is distributed
+along with
+[libstdcxx](https://github.com/gcc-mirror/gcc/tree/releases/gcc-13.3.0/libstdc%2B%2B-v3/python)
+itself.
+
+For now, you can download
+[this folder](https://github.com/gcc-mirror/gcc/tree/releases/gcc-13.3.0/libstdc%2B%2B-v3/python)
+and place it anywhere on your Haiku system. Then, use the `python` command to open a Python shell
+in GDB to import those scripts:
+
+```py
+(gdb) python
+>import sys
+>sys.path.insert(0, "/path/to/the/folder/you/saved")
+>from libstdcxx.v6.printers import register_libstdcxx_printers
+>register_libstdcxx_printers (None)
+>end
+(gdb)
+```
+
+After that, STL containers will be printed the way you expect:
+
+```
+(gdb) info locals
+event_count = 0
+wait_infos = std::vector of length 2, capacity 2 = {{object = 14, type = 0, events = 1}, {object = 242, type = 2, events = 1}}
+wait_for_ports = true
+__func__ = "operator()"
+(gdb)
+```
+
+It would be nice to package these pretty printers into `.hpkg`s and publish them to HaikuPorts one
+day. Another nice future idea is to build pretty printers for BeAPI types, such as `BLooper`s.
```

Comment:
If it comes with libstdc++, then we should distribute it with gcc_syslibs_devel? This sounds like it deserves a HaikuPorts ticket.

Reply:
Looks more to me like an optional component that deserves a separate package.

Reply:
Well either way we should open an issue at HaikuPorts with the details so we can solve the problem one way or another.

Reply:
https://github.com/haikuports/haikuports/issues/10826

Done.